{
  "cells": [
    {
      "cell_type": "code",
      "source": [
        "# --- Repo + path setup ---\n",
        "\n",
        "import os, sys\n",
        "\n",
        "# 1) Clone repo if it doesn't exist\n",
        "if not os.path.exists('/content/bdh_challenge_2025'):\n",
        "    %cd /content\n",
        "    !git clone https://github.com/arionandrei2000/bdh_challenge_2025.git\n",
        "\n",
        "# 2) Move into repo root\n",
        "%cd /content/bdh_challenge_2025\n",
        "print(\"PWD:\", os.getcwd())\n",
        "print(\"Repo contents:\", os.listdir())\n",
        "\n",
        "# 3) src is under notebooks/src, so add that parent folder to sys.path\n",
        "src_parent = \"/content/bdh_challenge_2025/notebooks\"\n",
        "if src_parent not in sys.path:\n",
        "    sys.path.append(src_parent)\n",
        "\n",
        "print(\"sys.path updated; ready to import src.\")\n"
      ],
      "metadata": {
        "id": "6QKChWUctneU",
        "outputId": "16bc7265-326f-41fb-b1bb-062349015391",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "id": "6QKChWUctneU",
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content\n",
            "Cloning into 'bdh_challenge_2025'...\n",
            "remote: Enumerating objects: 101, done.\u001b[K\n",
            "remote: Counting objects: 100% (101/101), done.\u001b[K\n",
            "remote: Compressing objects: 100% (92/92), done.\u001b[K\n",
            "remote: Total 101 (delta 16), reused 0 (delta 0), pack-reused 0 (from 0)\u001b[K\n",
            "Receiving objects: 100% (101/101), 40.98 KiB | 5.12 MiB/s, done.\n",
            "Resolving deltas: 100% (16/16), done.\n",
            "/content/bdh_challenge_2025\n",
            "PWD: /content/bdh_challenge_2025\n",
            "Repo contents: ['README.md', '.git', 'notebooks']\n",
            "sys.path updated; ready to import src.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 01) Dependencies\n",
        "\n",
        "# Core Python packages used in this notebook\n",
        "# - pandas / numpy: data manipulation\n",
        "# - pyarrow: for saving Parquet\n",
        "# - requests: HTTP requests to GDC API\n",
        "# - tqdm: progress bars\n",
        "!pip install pandas numpy pyarrow requests tqdm pyranges --quiet\n",
        "\n",
        "# Download GDC client\n",
        "!wget https://gdc.cancer.gov/files/public/file/gdc-client_v1.6.1_Ubuntu_x64.zip -O gdc.zip -q\n",
        "!unzip -o gdc.zip > /dev/null\n",
        "!chmod +x gdc-client\n",
        "\n",
        "print(\"Installed Python deps and downloaded gdc-client.\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 445
        },
        "id": "SeS8qFe8jDHd",
        "outputId": "e99a6ab2-ca02-4257-b975-50bfa066126a"
      },
      "id": "SeS8qFe8jDHd",
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[?25l     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.0/1.6 MB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\r\u001b[2K     \u001b[91m━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m\u001b[91m╸\u001b[0m\u001b[90m━━━━━━━━━━━\u001b[0m \u001b[32m1.1/1.6 MB\u001b[0m \u001b[31m34.1 MB/s\u001b[0m eta \u001b[36m0:00:01\u001b[0m\r\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m23.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Installing build dependencies ... \u001b[?25l\u001b[?25hcanceled\u001b[31mERROR: Operation cancelled by user\u001b[0m\u001b[31m\n",
            "\u001b[0m^C\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "KeyboardInterrupt",
          "evalue": "",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m/tmp/ipython-input-1039502551.py\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;31m# - tqdm: progress bars\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;31m# - pyranges: may be useful later for genomic ranges\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'pip install pandas numpy pyarrow requests tqdm pyranges --quiet'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;31m# Download GDC client\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/google/colab/_shell.py\u001b[0m in \u001b[0;36msystem\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    150\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    151\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mpip_warn\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 152\u001b[0;31m       \u001b[0m_pip\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprint_previous_import_warning\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutput\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    153\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    154\u001b[0m   \u001b[0;32mdef\u001b[0m \u001b[0m_send_error\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexc_content\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/google/colab/_pip.py\u001b[0m in \u001b[0;36mprint_previous_import_warning\u001b[0;34m(output)\u001b[0m\n\u001b[1;32m     54\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mprint_previous_import_warning\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutput\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     55\u001b[0m   \u001b[0;34m\"\"\"Prints a warning about previously imported packages.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 56\u001b[0;31m   \u001b[0mpackages\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_previously_imported_packages\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutput\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     57\u001b[0m   \u001b[0;32mif\u001b[0m \u001b[0mpackages\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     58\u001b[0m     \u001b[0;31m# display a list of packages using the colab-display-data mimetype, which\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/google/colab/_pip.py\u001b[0m in \u001b[0;36m_previously_imported_packages\u001b[0;34m(pip_output)\u001b[0m\n\u001b[1;32m     48\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0m_previously_imported_packages\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpip_output\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     49\u001b[0m   \u001b[0;34m\"\"\"List all previously imported packages from a pip install.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 50\u001b[0;31m   \u001b[0minstalled\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_extract_toplevel_packages\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpip_output\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     51\u001b[0m   \u001b[0;32mreturn\u001b[0m \u001b[0msorted\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minstalled\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mintersection\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmodules\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     52\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/google/colab/_pip.py\u001b[0m in \u001b[0;36m_extract_toplevel_packages\u001b[0;34m(pip_output)\u001b[0m\n\u001b[1;32m     37\u001b[0m   \u001b[0;34m\"\"\"Extract the list of toplevel packages associated with a pip install.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m   \u001b[0mtoplevel\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcollections\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdefaultdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 39\u001b[0;31m   \u001b[0;32mfor\u001b[0m \u001b[0mm\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mps\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mimportlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmetadata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpackages_distributions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     40\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mp\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mps\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     41\u001b[0m       \u001b[0mtoplevel\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mp\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mm\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/lib/python3.12/importlib/metadata/__init__.py\u001b[0m in \u001b[0;36mpackages_distributions\u001b[0;34m()\u001b[0m\n\u001b[1;32m    945\u001b[0m     \u001b[0mpkg_to_dist\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcollections\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdefaultdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    946\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mdist\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdistributions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 947\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0mpkg\u001b[0m \u001b[0;32min\u001b[0m \u001b[0m_top_level_declared\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdist\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0m_top_level_inferred\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdist\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    948\u001b[0m             \u001b[0mpkg_to_dist\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mpkg\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdist\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmetadata\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Name'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    949\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpkg_to_dist\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/lib/python3.12/importlib/metadata/__init__.py\u001b[0m in \u001b[0;36m_top_level_inferred\u001b[0;34m(dist)\u001b[0m\n\u001b[1;32m    957\u001b[0m     opt_names = {\n\u001b[1;32m    958\u001b[0m         \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparts\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparts\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0minspect\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetmodulename\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 959\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0mf\u001b[0m \u001b[0;32min\u001b[0m \u001b[0malways_iterable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdist\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfiles\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    960\u001b[0m     }\n\u001b[1;32m    961\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/lib/python3.12/importlib/metadata/__init__.py\u001b[0m in \u001b[0;36mfiles\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    498\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mpath\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlocate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexists\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpackage_paths\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    499\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 500\u001b[0;31m         return skip_missing_files(\n\u001b[0m\u001b[1;32m    501\u001b[0m             make_files(\n\u001b[1;32m    502\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_read_files_distinfo\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/lib/python3.12/importlib/metadata/_functools.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(param, *args, **kwargs)\u001b[0m\n\u001b[1;32m    100\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparam\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    101\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mparam\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 102\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparam\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    103\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    104\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/lib/python3.12/importlib/metadata/__init__.py\u001b[0m in \u001b[0;36mskip_missing_files\u001b[0;34m(package_paths)\u001b[0m\n\u001b[1;32m    496\u001b[0m         \u001b[0;34m@\u001b[0m\u001b[0mpass_none\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    497\u001b[0m         \u001b[0;32mdef\u001b[0m \u001b[0mskip_missing_files\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpackage_paths\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 498\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mpath\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlocate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexists\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpackage_paths\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    499\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    500\u001b[0m         return skip_missing_files(\n",
            "\u001b[0;32m/usr/lib/python3.12/importlib/metadata/__init__.py\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(path)\u001b[0m\n\u001b[1;32m    496\u001b[0m         \u001b[0;34m@\u001b[0m\u001b[0mpass_none\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    497\u001b[0m         \u001b[0;32mdef\u001b[0m \u001b[0mskip_missing_files\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpackage_paths\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 498\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mpath\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlocate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexists\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpackage_paths\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    499\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    500\u001b[0m         return skip_missing_files(\n",
            "\u001b[0;32m/usr/lib/python3.12/pathlib.py\u001b[0m in \u001b[0;36mexists\u001b[0;34m(self, follow_symlinks)\u001b[0m\n\u001b[1;32m    858\u001b[0m         \"\"\"\n\u001b[1;32m    859\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 860\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfollow_symlinks\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfollow_symlinks\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    861\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mOSError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    862\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0m_ignore_error\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/lib/python3.12/pathlib.py\u001b[0m in \u001b[0;36mstat\u001b[0;34m(self, follow_symlinks)\u001b[0m\n\u001b[1;32m    838\u001b[0m         \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0mdoes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    839\u001b[0m         \"\"\"\n\u001b[0;32m--> 840\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfollow_symlinks\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfollow_symlinks\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    841\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    842\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mlstat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "# 02) Mount GDrive\n",
        "# This lets us store large TCGA data in Drive so it persists across sessions\n",
        "\n",
        "from google.colab import drive\n",
        "drive.mount('/content/drive')\n",
        "\n",
        "print(\" Google Drive mounted at /content/drive\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PiI-cgscxSXN",
        "outputId": "90729b4a-5a40-4397-f2a5-497f1568298f"
      },
      "id": "PiI-cgscxSXN",
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n",
            " Google Drive mounted at /content/drive\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 03) Project Paths in Google Drive\n",
        "\n",
        "from pathlib import Path\n",
        "import os\n",
        "\n",
        "# Main project root on Google Drive\n",
        "PROJECT_ROOT = Path(\"/content/drive/MyDrive/bdh_challenge_2025_data\")\n",
        "PROJECT_ROOT.mkdir(exist_ok=True)\n",
        "\n",
        "# Data directories\n",
        "\n",
        "# Generic data directory\n",
        "DATA_DIR = PROJECT_ROOT / \"data\"\n",
        "DATA_DIR.mkdir(exist_ok=True)\n",
        "\n",
        "# Directory where raw TCGA STAR count files will be stored\n",
        "RNA_DIR = PROJECT_ROOT / \"tcga_rna\"\n",
        "RNA_DIR.mkdir(exist_ok=True)\n",
        "\n",
        "# Directory where processed matrices, tokenized data, embeddings\n",
        "PROCESSED_DIR = PROJECT_ROOT / \"processed\"\n",
        "PROCESSED_DIR.mkdir(exist_ok=True)\n",
        "\n",
        "print(\"Project root :\", PROJECT_ROOT)\n",
        "print(\"DATA_DIR     :\", DATA_DIR)\n",
        "print(\"RNA_DIR      :\", RNA_DIR)\n",
        "print(\"PROCESSED_DIR:\", PROCESSED_DIR)\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "E2yiIcDbj-mt",
        "outputId": "904b1cc1-7299-460d-80a6-b8e24f73e1d3"
      },
      "id": "E2yiIcDbj-mt",
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Project root : /content/drive/MyDrive/bdh_challenge_2025_data\n",
            "DATA_DIR     : /content/drive/MyDrive/bdh_challenge_2025_data/data\n",
            "RNA_DIR      : /content/drive/MyDrive/bdh_challenge_2025_data/tcga_rna\n",
            "PROCESSED_DIR: /content/drive/MyDrive/bdh_challenge_2025_data/processed\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 04) Imports and TCGA cohort\n",
        "\n",
        "import os\n",
        "import json\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import requests\n",
        "from tqdm import tqdm\n",
        "\n",
        "# We explicitly list the TCGA projects we will include. THIS CAN BE CHANGED!\n",
        "\n",
        "BULK_RNABERT_PROJECTS = [\n",
        "    \"TCGA-BRCA\",  # Breast invasive carcinoma\n",
        "    \"TCGA-BLCA\",  # Bladder urothelial carcinoma\n",
        "    \"TCGA-GBM\",   # Glioblastoma multiforme\n",
        "    \"TCGA-LGG\",   # Lower grade glioma\n",
        "    \"TCGA-LUAD\",  # Lung adenocarcinoma\n",
        "    \"TCGA-UCEC\",  # Uterine corpus endometrial carcinoma\n",
        "]\n",
        "\n",
        "print(\"Cohorts for this run:\")\n",
        "for p in BULK_RNABERT_PROJECTS:\n",
        "    print(\" -\", p)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "BRbiMaB9yAzh",
        "outputId": "325e95ec-7e28-4b5e-9bcc-e1fed88d54ba"
      },
      "id": "BRbiMaB9yAzh",
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cohorts for this run:\n",
            " - TCGA-BRCA\n",
            " - TCGA-BLCA\n",
            " - TCGA-GBM\n",
            " - TCGA-LGG\n",
            " - TCGA-LUAD\n",
            " - TCGA-UCEC\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "# 05) QUERY GDC FOR GENE EXPRESSION (STAR - COUNTS)\n",
        "\n",
        "# We ask GDC:\n",
        "#  - data_category: Transcriptome Profiling\n",
        "#  - data_type: Gene Expression Quantification\n",
        "#  - projects: the 6 TCGA projects above\n",
        "\n",
        "\n",
        "files_query = {\n",
        "    \"filters\": {\n",
        "        \"op\": \"and\",\n",
        "        \"content\": [\n",
        "            {\n",
        "                \"op\": \"in\",\n",
        "                \"content\": {\n",
        "                    \"field\": \"cases.project.project_id\",\n",
        "                    \"value\": BULK_RNABERT_PROJECTS,\n",
        "                },\n",
        "            },\n",
        "            {\n",
        "                \"op\": \"in\",\n",
        "                \"content\": {\n",
        "                    \"field\": \"data_category\",\n",
        "                    \"value\": [\"Transcriptome Profiling\"],\n",
        "                },\n",
        "            },\n",
        "            {\n",
        "                \"op\": \"in\",\n",
        "                \"content\": {\n",
        "                    \"field\": \"data_type\",\n",
        "                    \"value\": [\"Gene Expression Quantification\"],\n",
        "                },\n",
        "            },\n",
        "        ],\n",
        "    },\n",
        "    \"format\": \"JSON\",\n",
        "    \"size\": 20000,\n",
        "    \"fields\": (\n",
        "        \"file_id,file_name,\"\n",
        "        \"cases.submitter_id,cases.project.project_id,\"\n",
        "        \"data_category,data_type,analysis.workflow_type\"\n",
        "    ),\n",
        "}\n",
        "\n",
        "resp = requests.post(\n",
        "    \"https://api.gdc.cancer.gov/files\",\n",
        "    headers={\"Content-Type\": \"application/json\"},\n",
        "    data=json.dumps(files_query),\n",
        ")\n",
        "\n",
        "resp_json = resp.json()\n",
        "files_json = resp_json.get(\"data\", {}).get(\"hits\", [])\n",
        "\n",
        "print(\"HTTP status:\", resp.status_code)\n",
        "print(\"Number of files for selected projects:\", len(files_json))\n",
        "\n",
        "if len(files_json) == 0:\n",
        "    print(\" GDC returned 0 files. Pagination/debug:\")\n",
        "    print(json.dumps(resp_json.get(\"data\", {}).get(\"pagination\", {}), indent=2))\n",
        "    raise RuntimeError(\"GDC returned 0 files for this query. Adjust filters or check connection.\")\n",
        "\n",
        "# Convert JSON list to DataFrame\n",
        "files_df = pd.json_normalize(files_json)\n",
        "\n",
        "def _get_first_case(x):\n",
        "    \"\"\"Helper: for 'cases' field that is a list, return the first dict or {}.\"\"\"\n",
        "    if isinstance(x, list) and len(x) > 0 and isinstance(x[0], dict):\n",
        "        return x[0]\n",
        "    return {}\n",
        "\n",
        "# Extract submitter_id and project_id from nested 'cases' field\n",
        "files_df[\"submitter_id\"] = files_df[\"cases\"].apply(\n",
        "    lambda x: _get_first_case(x).get(\"submitter_id\", None)\n",
        ")\n",
        "files_df[\"project_id\"] = files_df[\"cases\"].apply(\n",
        "    lambda x: _get_first_case(x).get(\"project_id\", None)\n",
        ")\n",
        "\n",
        "print(\"Unique workflows in these files:\")\n",
        "print(files_df[\"analysis.workflow_type\"].value_counts())\n",
        "\n",
        "files_df[\n",
        "    [\"file_id\", \"file_name\", \"submitter_id\", \"project_id\",\n",
        "     \"data_category\", \"data_type\", \"analysis.workflow_type\"]\n",
        "].head()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 397
        },
        "id": "Zq9dx6PDpVn4",
        "outputId": "e77655b7-98d8-4546-fdbf-168e7ae72334"
      },
      "id": "Zq9dx6PDpVn4",
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "HTTP status: 200\n",
            "Number of files for selected projects: 3777\n",
            "Unique workflows in these files:\n",
            "analysis.workflow_type\n",
            "STAR - Counts    3777\n",
            "Name: count, dtype: int64\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                file_id  \\\n",
              "0  9dc09c86-c728-4bd9-b2b6-2d9962dad662   \n",
              "1  95668f0b-130d-44d4-94c0-ba7a4e7798e6   \n",
              "2  461fda5d-d6e6-4354-b035-c302cc43b03f   \n",
              "3  30ff778c-844b-4140-9025-7ab1938f10a9   \n",
              "4  427a04c9-9b48-49de-8a47-2adc4e1dd32a   \n",
              "\n",
              "                                           file_name  submitter_id project_id  \\\n",
              "0  d1f1743c-5fd9-4ae8-90c2-8c3e2d475d1b.rna_seq.a...  TCGA-EW-A2FS       None   \n",
              "1  6365a756-2e65-42cb-be4f-1f726915ca94.rna_seq.a...  TCGA-OL-A6VR       None   \n",
              "2  30285113-d411-475a-a8ed-1fb66be72f28.rna_seq.a...  TCGA-E9-A226       None   \n",
              "3  5167da8c-2b1c-4139-a2c4-355e9f07d0be.rna_seq.a...  TCGA-A8-A08H       None   \n",
              "4  fead73ce-2e66-4647-8b6c-b8bbdaaf30fe.rna_seq.a...  TCGA-D8-A27H       None   \n",
              "\n",
              "             data_category                       data_type  \\\n",
              "0  Transcriptome Profiling  Gene Expression Quantification   \n",
              "1  Transcriptome Profiling  Gene Expression Quantification   \n",
              "2  Transcriptome Profiling  Gene Expression Quantification   \n",
              "3  Transcriptome Profiling  Gene Expression Quantification   \n",
              "4  Transcriptome Profiling  Gene Expression Quantification   \n",
              "\n",
              "  analysis.workflow_type  \n",
              "0          STAR - Counts  \n",
              "1          STAR - Counts  \n",
              "2          STAR - Counts  \n",
              "3          STAR - Counts  \n",
              "4          STAR - Counts  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-32d6d613-e8a8-41ba-ad9f-8231f909b2d5\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>file_id</th>\n",
              "      <th>file_name</th>\n",
              "      <th>submitter_id</th>\n",
              "      <th>project_id</th>\n",
              "      <th>data_category</th>\n",
              "      <th>data_type</th>\n",
              "      <th>analysis.workflow_type</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>9dc09c86-c728-4bd9-b2b6-2d9962dad662</td>\n",
              "      <td>d1f1743c-5fd9-4ae8-90c2-8c3e2d475d1b.rna_seq.a...</td>\n",
              "      <td>TCGA-EW-A2FS</td>\n",
              "      <td>None</td>\n",
              "      <td>Transcriptome Profiling</td>\n",
              "      <td>Gene Expression Quantification</td>\n",
              "      <td>STAR - Counts</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>95668f0b-130d-44d4-94c0-ba7a4e7798e6</td>\n",
              "      <td>6365a756-2e65-42cb-be4f-1f726915ca94.rna_seq.a...</td>\n",
              "      <td>TCGA-OL-A6VR</td>\n",
              "      <td>None</td>\n",
              "      <td>Transcriptome Profiling</td>\n",
              "      <td>Gene Expression Quantification</td>\n",
              "      <td>STAR - Counts</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>461fda5d-d6e6-4354-b035-c302cc43b03f</td>\n",
              "      <td>30285113-d411-475a-a8ed-1fb66be72f28.rna_seq.a...</td>\n",
              "      <td>TCGA-E9-A226</td>\n",
              "      <td>None</td>\n",
              "      <td>Transcriptome Profiling</td>\n",
              "      <td>Gene Expression Quantification</td>\n",
              "      <td>STAR - Counts</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>30ff778c-844b-4140-9025-7ab1938f10a9</td>\n",
              "      <td>5167da8c-2b1c-4139-a2c4-355e9f07d0be.rna_seq.a...</td>\n",
              "      <td>TCGA-A8-A08H</td>\n",
              "      <td>None</td>\n",
              "      <td>Transcriptome Profiling</td>\n",
              "      <td>Gene Expression Quantification</td>\n",
              "      <td>STAR - Counts</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>427a04c9-9b48-49de-8a47-2adc4e1dd32a</td>\n",
              "      <td>fead73ce-2e66-4647-8b6c-b8bbdaaf30fe.rna_seq.a...</td>\n",
              "      <td>TCGA-D8-A27H</td>\n",
              "      <td>None</td>\n",
              "      <td>Transcriptome Profiling</td>\n",
              "      <td>Gene Expression Quantification</td>\n",
              "      <td>STAR - Counts</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-32d6d613-e8a8-41ba-ad9f-8231f909b2d5')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-32d6d613-e8a8-41ba-ad9f-8231f909b2d5 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-32d6d613-e8a8-41ba-ad9f-8231f909b2d5');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-1500db46-bd70-43dc-ade3-2e8fa28d7076\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-1500db46-bd70-43dc-ade3-2e8fa28d7076')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-1500db46-bd70-43dc-ade3-2e8fa28d7076 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "repr_error": "Out of range float values are not JSON compliant: nan"
            }
          },
          "metadata": {},
          "execution_count": 5
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 06) Create MANIFEST & DOWNLOAD star counts TO GDrive\n",
        "\n",
        "# GDC client expects a manifest file with an \"id\" column listing file_ids.\n",
        "\n",
        "manifest = files_df[[\"file_id\"]].rename(columns={\"file_id\": \"id\"})\n",
        "manifest_path = PROJECT_ROOT / \"manifest_star_counts.txt\"\n",
        "manifest.to_csv(manifest_path, sep=\"\\t\", index=False)\n",
        "\n",
        "print(\"Manifest written to:\", manifest_path)\n",
        "print(\"Total files listed:\", manifest.shape[0])\n",
        "\n",
        "# Download all files into RNA_DIR (in Google Drive).\n",
        "# gdc-client will skip files that are already complete.\n",
        "!./gdc-client download -m {manifest_path} -d {RNA_DIR}\n",
        "\n",
        "print(\" Download finished (check tcga_rna/ in Google Drive).\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3bAFfZGQp7In",
        "outputId": "273e36b7-a9e3-4f64-ce1a-9db856f53300"
      },
      "id": "3bAFfZGQp7In",
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Manifest written to: /content/drive/MyDrive/bdh_challenge_2025_data/manifest_star_counts.txt\n",
            "Total files listed: 3777\n",
            "/bin/bash: line 1: ./gdc-client: No such file or directory\n",
            " Download finished (check tcga_rna/ in Google Drive).\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 07) Getting the STAR counts matrix\n",
        "#\n",
        "# This version:\n",
        "#  - Only processes STAR gene count files (augmented_star_gene_counts.tsv)\n",
        "#  - Reads with header row\n",
        "#  - Drops N_* and __* summary rows\n",
        "#  - Works with partial downloads\n",
        "#  - Stores counts as float32 (half the RAM vs float64)\n",
        "\n",
        "id_to_sample = dict(zip(files_df[\"file_id\"], files_df[\"submitter_id\"]))\n",
        "print(\"Number of file_id → sample mappings:\", len(id_to_sample))\n",
        "\n",
        "all_files = []\n",
        "for root, dirs, files in os.walk(RNA_DIR):\n",
        "    for f in files:\n",
        "        # STAR gene count files usually contain 'rna_seq' and 'gene_counts'\n",
        "        if f.endswith(\".tsv\") and \"rna_seq\" in f and \"gene_counts\" in f:\n",
        "            all_files.append(os.path.join(root, f))\n",
        "\n",
        "print(\"Number of STAR gene count files on disk:\", len(all_files))\n",
        "\n",
        "MAX_FILES = None\n",
        "if MAX_FILES is not None:\n",
        "    all_files = all_files[:MAX_FILES]\n",
        "    print(f\"Using only first {len(all_files)} files.\")\n",
        "\n",
        "gene_index = None\n",
        "matrix = {}\n",
        "skipped = []\n",
        "\n",
        "for fpath in tqdm(all_files):\n",
        "    fname = os.path.basename(fpath)\n",
        "\n",
        "    # Folder name created by gdc-client is the file_id\n",
        "    file_id = os.path.basename(os.path.dirname(fpath))\n",
        "    sample = id_to_sample.get(file_id)\n",
        "    if sample is None:\n",
        "        skipped.append((fname, \"no sample mapping\"))\n",
        "        continue\n",
        "\n",
        "    # Read gene counts table; ignore comment lines starting with '#'\n",
        "    df = pd.read_csv(fpath, sep=\"\\t\", comment=\"#\")\n",
        "\n",
        "    # Identify gene ID column\n",
        "    if \"gene_id\" in df.columns:\n",
        "        gene_col = \"gene_id\"\n",
        "    elif \"Geneid\" in df.columns:\n",
        "        gene_col = \"Geneid\"\n",
        "    else:\n",
        "        skipped.append((fname, f\"no gene_id column, cols = {df.columns.tolist()}\"))\n",
        "        continue\n",
        "\n",
        "    # Identify counts column\n",
        "    if \"unstranded\" in df.columns:\n",
        "        counts_col = \"unstranded\"\n",
        "    else:\n",
        "        # pick the first non-gene_id/gene_name column as fallback\n",
        "        candidate_cols = [c for c in df.columns if c not in [gene_col, \"gene_name\"]]\n",
        "        if not candidate_cols:\n",
        "            skipped.append((fname, \"no suitable count column\"))\n",
        "            continue\n",
        "        counts_col = candidate_cols[0]\n",
        "\n",
        "    # Drop technical summary rows: N_unmapped, N_multimapping, __no_feature, etc.\n",
        "    df = df[~df[gene_col].astype(str).str.startswith(\"N_\")]\n",
        "    df = df[~df[gene_col].astype(str).str.startswith(\"__\")]\n",
        "    df = df.reset_index(drop=True)\n",
        "\n",
        "    # Initialize or check gene ordering\n",
        "    if gene_index is None:\n",
        "        gene_index = df[gene_col].values\n",
        "    else:\n",
        "        if not np.array_equal(gene_index, df[gene_col].values):\n",
        "            raise ValueError(f\"Gene order mismatch in file: {fpath}\")\n",
        "\n",
        "    # Extract counts as float32 (saves RAM vs float64)\n",
        "    counts = pd.to_numeric(df[counts_col], errors=\"raise\").astype(\"float32\").values\n",
        "    matrix[sample] = counts\n",
        "\n",
        "# Build counts matrix (genes x samples) and store as float32\n",
        "expr_counts = pd.DataFrame(matrix, index=gene_index).astype(\"float32\")\n",
        "\n",
        "print(\"Counts matrix shape (genes x samples):\", expr_counts.shape)\n",
        "display(expr_counts.iloc[:5, :5])\n",
        "\n",
        "print(\"Skipped files:\", len(skipped))\n",
        "if skipped:\n",
        "    print(skipped[:5])\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 293
        },
        "id": "uX3_6RQ1210i",
        "outputId": "6f71b2d1-cf58-440c-877d-82379435036b"
      },
      "id": "uX3_6RQ1210i",
      "execution_count": null,
      "outputs": [
        {
          "metadata": {
            "tags": null
          },
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Number of file_id → sample mappings: 3777\n",
            "Number of STAR gene count files on disk: 3777\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 3777/3777 [52:40<00:00,  1.20it/s]\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Counts matrix shape (genes x samples): (60660, 3385)\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "                    TCGA-P5-A5EY  TCGA-AR-A1AW  TCGA-E9-A1N5  TCGA-97-7941  \\\n",
              "ENSG00000000003.15        2377.0        2273.0        2036.0       11528.0   \n",
              "ENSG00000000005.6            8.0          13.0          39.0          13.0   \n",
              "ENSG00000000419.13         649.0        2077.0         703.0        1547.0   \n",
              "ENSG00000000457.14         331.0        1459.0         764.0        1257.0   \n",
              "ENSG00000000460.17          79.0         726.0         251.0         232.0   \n",
              "\n",
              "                    TCGA-93-7347  \n",
              "ENSG00000000003.15        1381.0  \n",
              "ENSG00000000005.6            5.0  \n",
              "ENSG00000000419.13         998.0  \n",
              "ENSG00000000457.14         715.0  \n",
              "ENSG00000000460.17         195.0  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-6a2fddd3-48de-4e73-a969-70b6de8d9cec\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>TCGA-P5-A5EY</th>\n",
              "      <th>TCGA-AR-A1AW</th>\n",
              "      <th>TCGA-E9-A1N5</th>\n",
              "      <th>TCGA-97-7941</th>\n",
              "      <th>TCGA-93-7347</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>ENSG00000000003.15</th>\n",
              "      <td>2377.0</td>\n",
              "      <td>2273.0</td>\n",
              "      <td>2036.0</td>\n",
              "      <td>11528.0</td>\n",
              "      <td>1381.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSG00000000005.6</th>\n",
              "      <td>8.0</td>\n",
              "      <td>13.0</td>\n",
              "      <td>39.0</td>\n",
              "      <td>13.0</td>\n",
              "      <td>5.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSG00000000419.13</th>\n",
              "      <td>649.0</td>\n",
              "      <td>2077.0</td>\n",
              "      <td>703.0</td>\n",
              "      <td>1547.0</td>\n",
              "      <td>998.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSG00000000457.14</th>\n",
              "      <td>331.0</td>\n",
              "      <td>1459.0</td>\n",
              "      <td>764.0</td>\n",
              "      <td>1257.0</td>\n",
              "      <td>715.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSG00000000460.17</th>\n",
              "      <td>79.0</td>\n",
              "      <td>726.0</td>\n",
              "      <td>251.0</td>\n",
              "      <td>232.0</td>\n",
              "      <td>195.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6a2fddd3-48de-4e73-a969-70b6de8d9cec')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-6a2fddd3-48de-4e73-a969-70b6de8d9cec button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-6a2fddd3-48de-4e73-a969-70b6de8d9cec');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-42d3ff5c-f040-4691-8143-e2d7ff4a8ee5\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-42d3ff5c-f040-4691-8143-e2d7ff4a8ee5')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-42d3ff5c-f040-4691-8143-e2d7ff4a8ee5 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"    print(skipped[:5])\",\n  \"rows\": 5,\n  \"fields\": [\n    {\n      \"column\": \"TCGA-P5-A5EY\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          8.0,\n          79.0,\n          649.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"TCGA-AR-A1AW\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          13.0,\n          726.0,\n          2077.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"TCGA-E9-A1N5\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          39.0,\n          251.0,\n          703.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"TCGA-97-7941\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          13.0,\n          232.0,\n          1547.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"TCGA-93-7347\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          5.0,\n          195.0,\n          998.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Skipped files: 0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 08) Doing TPM-like normalization on Counts\n",
        "# We approximatd TPM by:\n",
        "#   TPM_like = (counts / sum(counts)) * 1e6 per sample.\n",
        "# This keeps per-sample library size comparable across samples.\n",
        "\n",
        "# Very close to using FPKM→TPM for a lot of downstream modeling use cases. LET'S double check this! Why?\n",
        "\n",
        "def counts_to_tpm_like(column: pd.Series) -> pd.Series:\n",
        "    total = column.sum()\n",
        "    if total == 0:\n",
        "        return column\n",
        "    return (column / total) * 1e6\n",
        "\n",
        "expr_tpm = expr_counts.apply(counts_to_tpm_like, axis=0)\n",
        "print(\"TPM-like matrix shape (genes x samples):\", expr_tpm.shape)\n",
        "expr_tpm.iloc[:5, :5]\n"
      ],
      "metadata": {
        "id": "PnbLAhf84bvP",
        "outputId": "ba441855-7375-43a7-b066-85ebbbf052df",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 224
        }
      },
      "id": "PnbLAhf84bvP",
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "TPM-like matrix shape (genes x samples): (60660, 3385)\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                    TCGA-P5-A5EY  TCGA-AR-A1AW  TCGA-E9-A1N5  TCGA-97-7941  \\\n",
              "ENSG00000000003.15     47.031635     39.358894     69.550377    178.855652   \n",
              "ENSG00000000005.6       0.158289      0.225106      1.332252      0.201694   \n",
              "ENSG00000000419.13     12.841199     35.964989     24.014692     24.001535   \n",
              "ENSG00000000457.14      6.549210     25.263803     26.098471     19.502218   \n",
              "ENSG00000000460.17      1.563104     12.571296      8.574236      3.599455   \n",
              "\n",
              "                    TCGA-93-7347  \n",
              "ENSG00000000003.15     23.861710  \n",
              "ENSG00000000005.6       0.086393  \n",
              "ENSG00000000419.13     17.244017  \n",
              "ENSG00000000457.14     12.354180  \n",
              "ENSG00000000460.17      3.369322  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-6d48525d-9e26-4c04-9fa9-304c39894f1d\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>TCGA-P5-A5EY</th>\n",
              "      <th>TCGA-AR-A1AW</th>\n",
              "      <th>TCGA-E9-A1N5</th>\n",
              "      <th>TCGA-97-7941</th>\n",
              "      <th>TCGA-93-7347</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>ENSG00000000003.15</th>\n",
              "      <td>47.031635</td>\n",
              "      <td>39.358894</td>\n",
              "      <td>69.550377</td>\n",
              "      <td>178.855652</td>\n",
              "      <td>23.861710</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSG00000000005.6</th>\n",
              "      <td>0.158289</td>\n",
              "      <td>0.225106</td>\n",
              "      <td>1.332252</td>\n",
              "      <td>0.201694</td>\n",
              "      <td>0.086393</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSG00000000419.13</th>\n",
              "      <td>12.841199</td>\n",
              "      <td>35.964989</td>\n",
              "      <td>24.014692</td>\n",
              "      <td>24.001535</td>\n",
              "      <td>17.244017</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSG00000000457.14</th>\n",
              "      <td>6.549210</td>\n",
              "      <td>25.263803</td>\n",
              "      <td>26.098471</td>\n",
              "      <td>19.502218</td>\n",
              "      <td>12.354180</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSG00000000460.17</th>\n",
              "      <td>1.563104</td>\n",
              "      <td>12.571296</td>\n",
              "      <td>8.574236</td>\n",
              "      <td>3.599455</td>\n",
              "      <td>3.369322</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6d48525d-9e26-4c04-9fa9-304c39894f1d')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-6d48525d-9e26-4c04-9fa9-304c39894f1d button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-6d48525d-9e26-4c04-9fa9-304c39894f1d');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-8c197e7e-e59f-4a56-a530-9be163449bce\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-8c197e7e-e59f-4a56-a530-9be163449bce')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-8c197e7e-e59f-4a56-a530-9be163449bce button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"expr_tpm\",\n  \"rows\": 5,\n  \"fields\": [\n    {\n      \"column\": \"TCGA-P5-A5EY\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          0.15828904509544373,\n          1.563104271888733,\n          12.841198921203613\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"TCGA-AR-A1AW\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          0.22510585188865662,\n          12.571295738220215,\n          35.964988708496094\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"TCGA-E9-A1N5\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          1.332251787185669,\n          8.574235916137695,\n          24.014692306518555\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"TCGA-97-7941\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          0.2016935795545578,\n          3.599454641342163,\n          24.001535415649414\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"TCGA-93-7347\",\n      \"properties\": {\n        \"dtype\": \"float32\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          0.08639286458492279,\n          3.369321823120117,\n          17.244016647338867\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 09) Clean genes, reorient to samples x genes, log-transform, and save\n",
        "#\n",
        "# Starting point:\n",
        "#   - expr_counts : (genes x samples), float32\n",
        "#   - expr_tpm    : (genes x samples), TPM-like\n",
        "#\n",
        "# This cell:\n",
        "#   (1) Strips Ensembl version suffixes (ENSG...15 -> ENSG...)\n",
        "#   (2) Removes duplicates in BOTH matrices\n",
        "#   (3) Intersects gene sets so counts & TPM have identical genes\n",
        "#   (4) Reorients to (samples x genes)\n",
        "#   (5) Applies log10(1 + TPM-like)\n",
        "#   (6) Saves outputs\n",
        "\n",
        "import gc\n",
        "import numpy as np\n",
        "import os\n",
        "\n",
        "# 1) Strip Ensembl version suffix from gene IDs\n",
        "expr_tpm.index    = expr_tpm.index.to_series().str.split(\".\").str[0]\n",
        "expr_counts.index = expr_counts.index.to_series().str.split(\".\").str[0]\n",
        "\n",
        "\n",
        "# 2) Remove duplicate genes IN BOTH MATRICES\n",
        "expr_tpm    = expr_tpm[~expr_tpm.index.duplicated(keep=\"first\")]\n",
        "expr_counts = expr_counts[~expr_counts.index.duplicated(keep=\"first\")]\n",
        "\n",
        "# 3) Intersect genes so both matrices have identical gene sets\n",
        "common_genes = expr_tpm.index.intersection(expr_counts.index)\n",
        "expr_tpm    = expr_tpm.loc[common_genes].sort_index()\n",
        "expr_counts = expr_counts.loc[common_genes].sort_index()\n",
        "\n",
        "print(\"After strip + dedupe + intersect:\")\n",
        "print(\"  expr_counts shape:\", expr_counts.shape)\n",
        "print(\"  expr_tpm    shape:\", expr_tpm.shape)\n",
        "\n",
        "# 4) Reorient to (samples x genes)\n",
        "expr_counts_sxg = expr_counts.T.astype(\"float32\").copy()\n",
        "expr_tpm_sxg    = expr_tpm.T.astype(\"float32\").copy()\n",
        "\n",
        "expr_counts_sxg.index.name = \"sample_id\"\n",
        "expr_tpm_sxg.index.name    = \"sample_id\"\n",
        "\n",
        "print(\"Counts   (samples x genes):\", expr_counts_sxg.shape)\n",
        "print(\"TPM-like (samples x genes):\", expr_tpm_sxg.shape)\n",
        "\n",
        "gc.collect()\n",
        "\n",
        "# 5) Log-transform: log10(1 + TPM-like)\n",
        "expr_tpm_log = np.log10(1.0 + expr_tpm_sxg.astype(\"float32\"))\n",
        "print(\"Log10(1 + TPM-like) shape:\", expr_tpm_log.shape)\n",
        "\n",
        "# 6) Save matrices to disk (into Google Drive)\n",
        "from pathlib import Path\n",
        "\n",
        "# Use the PROCESSED_DIR defined in the paths cell\n",
        "OUT_DIR = PROCESSED_DIR  # this is a Path object pointing to .../bdh_challenge_2025_data/processed\n",
        "OUT_DIR.mkdir(parents=True, exist_ok=True)\n",
        "\n",
        "expr_counts_sxg.to_parquet(OUT_DIR / \"tcga_star_counts_sxg.parquet\")\n",
        "expr_tpm_sxg.to_parquet(OUT_DIR / \"tcga_tpm_like_sxg.parquet\")\n",
        "expr_tpm_log.to_parquet(OUT_DIR / \"tcga_tpm_like_log10_sxg.parquet\")\n",
        "\n",
        "print(\"Saved processed matrices to:\", OUT_DIR)\n",
        "\n",
        "# Optional: verify files\n",
        "print(\"Processed directory contents:\")\n",
        "for p in OUT_DIR.iterdir():\n",
        "    print(\"  -\", p.name)\n"
      ],
      "metadata": {
        "id": "Qm7WosVAJeUl",
        "outputId": "f8b87cd5-547a-481a-d72c-1e2def8ce53f",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "id": "Qm7WosVAJeUl",
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "After strip + dedupe + intersect:\n",
            "  expr_counts shape: (60616, 3385)\n",
            "  expr_tpm    shape: (60616, 3385)\n",
            "Counts   (samples x genes): (3385, 60616)\n",
            "TPM-like (samples x genes): (3385, 60616)\n",
            "Log10(1 + TPM-like) shape: (3385, 60616)\n",
            "Saved processed matrices to: /content/drive/MyDrive/bdh_challenge_2025_data/processed\n",
            "Processed directory contents:\n",
            "  - tcga_star_counts_sxg.parquet\n",
            "  - tcga_tpm_like_sxg.parquet\n",
            "  - tcga_tpm_like_log10_sxg.parquet\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 10) Align our TCGA log-TPM matrix to BulkRNABert gene list (common_gene_id.txt)\n",
        "\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "from pathlib import Path\n",
        "\n",
        "GENE_LIST_PATH = DATA_DIR / \"common_gene_id.txt\"\n",
        "TPM_LOG_PATH   = PROCESSED_DIR / \"tcga_tpm_like_log10_sxg.parquet\"\n",
        "\n",
        "print(\"Gene list path:\", GENE_LIST_PATH)\n",
        "print(\"Log TPM matrix path:\", TPM_LOG_PATH)\n",
        "\n",
        "# 1) Load authors' gene list\n",
        "# File is one gene ID per line\n",
        "with open(GENE_LIST_PATH) as f:\n",
        "    author_genes = [line.strip() for line in f if line.strip()]\n",
        "\n",
        "author_genes = pd.Index(author_genes, name=\"gene_id\")\n",
        "print(\"Genes in authors' list:\", len(author_genes))\n",
        "print(author_genes[:5])\n",
        "\n",
        "# 2) Load our processed log TPM matrix (samples x genes)\n",
        "expr_tpm_log = pd.read_parquet(TPM_LOG_PATH)\n",
        "print(\"Our TCGA log-TPM matrix shape (samples x genes):\", expr_tpm_log.shape)\n",
        "\n",
        "# Make sure columns are strings (gene IDs)\n",
        "expr_tpm_log.columns = expr_tpm_log.columns.astype(str)\n",
        "\n",
        "# 3) Check overlaps between our genes and authors' genes\n",
        "our_genes = pd.Index(expr_tpm_log.columns, name=\"gene_id\")\n",
        "common = author_genes.intersection(our_genes)\n",
        "missing_in_ours = author_genes.difference(our_genes)\n",
        "extra_in_ours = our_genes.difference(author_genes)\n",
        "\n",
        "print(\"\\n=== Gene set diagnostics ===\")\n",
        "print(\"Genes in our matrix        :\", len(our_genes))\n",
        "print(\"Genes in authors' list     :\", len(author_genes))\n",
        "print(\"Common genes               :\", len(common))\n",
        "print(\"Authors' genes missing in ours:\", len(missing_in_ours))\n",
        "print(\"Our genes not in authors' list:\", len(extra_in_ours))\n",
        "\n",
        "if len(common) == 0:\n",
        "    raise ValueError(\"No overlap between TCGA genes and author gene list – check ID format!\")\n",
        "\n",
        "print(\"\\nExample authors' genes missing in our matrix:\", list(missing_in_ours[:10]))\n",
        "\n",
        "# 4) Reindex to authors' order (BulkRNABert canonical order)\n",
        "# This will create columns for ALL author_genes, preserving order.\n",
        "# Genes that we don't have will be NaN.\n",
        "expr_aligned = expr_tpm_log.reindex(columns=author_genes)\n",
        "\n",
        "print(\"\\nAligned matrix shape (samples x authors' genes):\", expr_aligned.shape)\n",
        "\n",
        "# 5) Handle missing values & basic sanity checks\n",
        "nan_count = expr_aligned.isna().sum().sum()\n",
        "print(\"Total NaN values in aligned matrix:\", nan_count)\n",
        "\n",
        "# Strategy: fill missing genes with 0.0 (log10(1+TPM)=0 => TPM=0)\n",
        "expr_aligned = expr_aligned.fillna(0.0).astype(\"float32\")\n",
        "\n",
        "# Quick distribution check\n",
        "print(\"\\nAligned matrix summary (first 5 genes):\")\n",
        "print(expr_aligned.iloc[:, :5].describe().T)\n",
        "\n",
        "# 6) Save aligned matrix\n",
        "ALIGNED_PATH = PROCESSED_DIR / \"tcga_tpm_like_log10_bulkrnabert_aligned.parquet\"\n",
        "expr_aligned.to_parquet(ALIGNED_PATH)\n",
        "\n",
        "print(\"\\nSaved aligned matrix to:\", ALIGNED_PATH)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-x3iPN1PB1sl",
        "outputId": "4bb56470-77f0-416d-fdfd-f744aa2253d2"
      },
      "id": "-x3iPN1PB1sl",
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Gene list path: /content/drive/MyDrive/bdh_challenge_2025_data/data/common_gene_id.txt\n",
            "Log TPM matrix path: /content/drive/MyDrive/bdh_challenge_2025_data/processed/tcga_tpm_like_log10_sxg.parquet\n",
            "Genes in authors' list: 19062\n",
            "Index(['ENSG00000000003', 'ENSG00000000005', 'ENSG00000000419',\n",
            "       'ENSG00000000457', 'ENSG00000000460'],\n",
            "      dtype='object', name='gene_id')\n",
            "Our TCGA log-TPM matrix shape (samples x genes): (3385, 60616)\n",
            "\n",
            "=== Gene set diagnostics ===\n",
            "Genes in our matrix        : 60616\n",
            "Genes in authors' list     : 19062\n",
            "Common genes               : 19062\n",
            "Authors' genes missing in ours: 0\n",
            "Our genes not in authors' list: 41554\n",
            "\n",
            "Example authors' genes missing in our matrix: []\n",
            "\n",
            "Aligned matrix shape (samples x authors' genes): (3385, 19062)\n",
            "Total NaN values in aligned matrix: 0\n",
            "\n",
            "Aligned matrix summary (first 5 genes):\n",
            "                  count      mean       std       min       25%       50%  \\\n",
            "gene_id                                                                     \n",
            "ENSG00000000003  3385.0  1.794134  0.296457  0.444586  1.627818  1.802887   \n",
            "ENSG00000000005  3385.0  0.111150  0.238335  0.000000  0.008045  0.032830   \n",
            "ENSG00000000419  3385.0  1.502516  0.203080  0.304798  1.364327  1.500049   \n",
            "ENSG00000000457  3385.0  1.248408  0.214124  0.520206  1.093054  1.226594   \n",
            "ENSG00000000460  3385.0  0.961392  0.259730  0.101108  0.781082  0.959704   \n",
            "\n",
            "                      75%       max  \n",
            "gene_id                              \n",
            "ENSG00000000003  1.985771  3.005660  \n",
            "ENSG00000000005  0.093935  2.575137  \n",
            "ENSG00000000419  1.629153  2.512778  \n",
            "ENSG00000000457  1.395588  2.107116  \n",
            "ENSG00000000460  1.132621  2.116103  \n",
            "\n",
            "Saved aligned matrix to: /content/drive/MyDrive/bdh_challenge_2025_data/processed/tcga_tpm_like_log10_bulkrnabert_aligned.parquet\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 11) Download TCGA clinical data (cases endpoint) for the selected cohorts\n",
        "\n",
        "import json\n",
        "import requests\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "from tqdm import tqdm\n",
        "\n",
        "print(\"Projects:\", BULK_RNABERT_PROJECTS)\n",
        "\n",
        "cases_query = {\n",
        "    \"filters\": {\n",
        "        \"op\": \"in\",\n",
        "        \"content\": {\n",
        "            \"field\": \"project.project_id\",\n",
        "            \"value\": BULK_RNABERT_PROJECTS,\n",
        "        },\n",
        "    },\n",
        "    \"format\": \"JSON\",\n",
        "    \"size\": 10000,  # enough for all cases in these 6 cohorts\n",
        "    \"fields\": \",\".join([\n",
        "        \"case_id\",\n",
        "        \"submitter_id\",\n",
        "        \"project.project_id\",\n",
        "        # demographics\n",
        "        \"demographic.gender\",\n",
        "        \"demographic.race\",\n",
        "        \"demographic.ethnicity\",\n",
        "        # diagnoses (may be multiple per case)\n",
        "        \"diagnoses.vital_status\",\n",
        "        \"diagnoses.days_to_death\",\n",
        "        \"diagnoses.days_to_last_follow_up\",\n",
        "        \"diagnoses.days_to_last_known_alive\",\n",
        "        \"diagnoses.diagnosis_is_primary_disease\",\n",
        "        \"diagnoses.age_at_diagnosis\",\n",
        "        \"diagnoses.primary_diagnosis\",\n",
        "        \"diagnoses.tumor_stage\",\n",
        "        \"diagnoses.morphology\",\n",
        "    ]),\n",
        "}\n",
        "\n",
        "resp = requests.post(\n",
        "    \"https://api.gdc.cancer.gov/cases\",\n",
        "    headers={\"Content-Type\": \"application/json\"},\n",
        "    data=json.dumps(cases_query),\n",
        ")\n",
        "\n",
        "print(\"HTTP status:\", resp.status_code)\n",
        "cases_json = resp.json()\n",
        "\n",
        "cases_hits = cases_json.get(\"data\", {}).get(\"hits\", [])\n",
        "print(\"Number of case records returned:\", len(cases_hits))\n",
        "\n",
        "if len(cases_hits) == 0:\n",
        "    raise RuntimeError(\"No clinical cases returned from GDC. Check query or network.\")\n"
      ],
      "metadata": {
        "id": "pjuS9Go-3tOy",
        "outputId": "6a8dde25-336c-4e9b-a748-9a49a5710e00",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "id": "pjuS9Go-3tOy",
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Projects: ['TCGA-BRCA', 'TCGA-BLCA', 'TCGA-GBM', 'TCGA-LGG', 'TCGA-LUAD', 'TCGA-UCEC']\n",
            "HTTP status: 200\n",
            "Number of case records returned: 3788\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def _safe_get(d, *keys, default=None):\n",
        "    \"\"\"Nested dict .get helper: _safe_get(d, 'a', 'b') -> d['a']['b'] if present.\"\"\"\n",
        "    cur = d\n",
        "    for k in keys:\n",
        "        if cur is None:\n",
        "            return default\n",
        "        if isinstance(cur, dict):\n",
        "            cur = cur.get(k, default)\n",
        "        else:\n",
        "            return default\n",
        "    return cur if cur is not None else default\n",
        "\n",
        "\n",
        "def _pick_primary_diagnosis(diags):\n",
        "    \"\"\"\n",
        "    diags: list of diagnosis dicts (may be empty).\n",
        "    Preference:\n",
        "      - any with diagnosis_is_primary_disease == True\n",
        "      - else the first one\n",
        "    \"\"\"\n",
        "    if not diags:\n",
        "        return None\n",
        "    for d in diags:\n",
        "        if d.get(\"diagnosis_is_primary_disease\") is True:\n",
        "            return d\n",
        "    return diags[0]\n",
        "\n",
        "\n",
        "rows = []\n",
        "for c in tqdm(cases_hits, desc=\"Building clinical rows\"):\n",
        "    submitter = c.get(\"submitter_id\")    # e.g. TCGA-P5-A5EY\n",
        "    if submitter is None:\n",
        "        continue\n",
        "\n",
        "    project_id = _safe_get(c, \"project\", \"project_id\")\n",
        "    cohort = None\n",
        "    if project_id is not None:\n",
        "        # TCGA-BRCA -> BRCA\n",
        "        cohort = project_id.replace(\"TCGA-\", \"\")\n",
        "\n",
        "    demo = c.get(\"demographic\") or {}\n",
        "    gender = demo.get(\"gender\")\n",
        "    race = demo.get(\"race\")\n",
        "    ethnicity = demo.get(\"ethnicity\")\n",
        "\n",
        "    diags = c.get(\"diagnoses\") or []\n",
        "    primary = _pick_primary_diagnosis(diags)\n",
        "\n",
        "    vital_status = None\n",
        "    days_to_death = None\n",
        "    days_to_followup = None\n",
        "    age_at_diag = None\n",
        "    tumor_stage = None\n",
        "    primary_dx = None\n",
        "    morphology = None\n",
        "\n",
        "    if primary is not None:\n",
        "        vital_status = primary.get(\"vital_status\")\n",
        "        days_to_death = primary.get(\"days_to_death\")\n",
        "        days_to_followup = (\n",
        "            primary.get(\"days_to_last_follow_up\")\n",
        "            if primary.get(\"days_to_last_follow_up\") is not None\n",
        "            else primary.get(\"days_to_last_known_alive\")\n",
        "        )\n",
        "        age_at_diag = primary.get(\"age_at_diagnosis\")\n",
        "        tumor_stage = primary.get(\"tumor_stage\")\n",
        "        primary_dx = primary.get(\"primary_diagnosis\")\n",
        "        morphology = primary.get(\"morphology\")\n",
        "\n",
        "    # Compute survival_time and event\n",
        "    survival_time = None\n",
        "    event = None\n",
        "\n",
        "    # Use days_to_death if dead, else days_to_last_follow_up/known_alive\n",
        "    if vital_status is not None:\n",
        "        if vital_status.upper() == \"DEAD\" and days_to_death is not None:\n",
        "            event = 1\n",
        "            survival_time = days_to_death\n",
        "        elif days_to_followup is not None:\n",
        "            event = 0\n",
        "            survival_time = days_to_followup\n",
        "\n",
        "    row = {\n",
        "        \"sample_id\": submitter,\n",
        "        \"project_id\": project_id,\n",
        "        \"cohort\": cohort,\n",
        "        \"gender\": gender,\n",
        "        \"race\": race,\n",
        "        \"ethnicity\": ethnicity,\n",
        "        \"vital_status\": vital_status,\n",
        "        \"days_to_death\": days_to_death,\n",
        "        \"days_to_last_followup\": days_to_followup,\n",
        "        \"age_at_diagnosis\": age_at_diag,\n",
        "        \"tumor_stage\": tumor_stage,\n",
        "        \"primary_diagnosis\": primary_dx,\n",
        "        \"morphology\": morphology,\n",
        "        \"survival_time\": survival_time,\n",
        "        \"event\": event,\n",
        "    }\n",
        "    rows.append(row)\n",
        "\n",
        "clinical_df = pd.DataFrame(rows)\n",
        "print(\"Raw clinical table shape:\", clinical_df.shape)\n",
        "clinical_df.head()\n"
      ],
      "metadata": {
        "id": "Ctr9Za1S4J91",
        "outputId": "a3a6e754-55e7-4b12-81ae-f87c37218789",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 434
        }
      },
      "id": "Ctr9Za1S4J91",
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Building clinical rows: 100%|██████████| 3788/3788 [00:00<00:00, 193624.15it/s]"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Raw clinical table shape: (3788, 15)\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "      sample_id project_id cohort  gender                       race  \\\n",
              "0  TCGA-E2-A107  TCGA-BRCA   BRCA  female                      white   \n",
              "1  TCGA-AC-A62X  TCGA-BRCA   BRCA  female  black or african american   \n",
              "2  TCGA-E2-A156  TCGA-BRCA   BRCA  female                      white   \n",
              "3  TCGA-A7-A56D  TCGA-BRCA   BRCA  female  black or african american   \n",
              "4  TCGA-AR-A0U1  TCGA-BRCA   BRCA  female                      white   \n",
              "\n",
              "                ethnicity vital_status days_to_death  days_to_last_followup  \\\n",
              "0  not hispanic or latino         None          None                 1047.0   \n",
              "1  not hispanic or latino         None          None                  417.0   \n",
              "2  not hispanic or latino         None          None                  726.0   \n",
              "3  not hispanic or latino         None          None                  448.0   \n",
              "4  not hispanic or latino         None          None                 4052.0   \n",
              "\n",
              "   age_at_diagnosis tumor_stage                 primary_diagnosis morphology  \\\n",
              "0           20065.0        None            Lobular carcinoma, NOS     8520/3   \n",
              "1           26627.0        None  Infiltrating duct carcinoma, NOS     8500/3   \n",
              "2           22422.0        None  Infiltrating duct carcinoma, NOS     8500/3   \n",
              "3           30783.0        None  Infiltrating duct carcinoma, NOS     8500/3   \n",
              "4           13467.0        None  Infiltrating duct carcinoma, NOS     8500/3   \n",
              "\n",
              "  survival_time event  \n",
              "0          None  None  \n",
              "1          None  None  \n",
              "2          None  None  \n",
              "3          None  None  \n",
              "4          None  None  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-be5db493-b001-4acf-8280-36915682f7b6\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>sample_id</th>\n",
              "      <th>project_id</th>\n",
              "      <th>cohort</th>\n",
              "      <th>gender</th>\n",
              "      <th>race</th>\n",
              "      <th>ethnicity</th>\n",
              "      <th>vital_status</th>\n",
              "      <th>days_to_death</th>\n",
              "      <th>days_to_last_followup</th>\n",
              "      <th>age_at_diagnosis</th>\n",
              "      <th>tumor_stage</th>\n",
              "      <th>primary_diagnosis</th>\n",
              "      <th>morphology</th>\n",
              "      <th>survival_time</th>\n",
              "      <th>event</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>TCGA-E2-A107</td>\n",
              "      <td>TCGA-BRCA</td>\n",
              "      <td>BRCA</td>\n",
              "      <td>female</td>\n",
              "      <td>white</td>\n",
              "      <td>not hispanic or latino</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>1047.0</td>\n",
              "      <td>20065.0</td>\n",
              "      <td>None</td>\n",
              "      <td>Lobular carcinoma, NOS</td>\n",
              "      <td>8520/3</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>TCGA-AC-A62X</td>\n",
              "      <td>TCGA-BRCA</td>\n",
              "      <td>BRCA</td>\n",
              "      <td>female</td>\n",
              "      <td>black or african american</td>\n",
              "      <td>not hispanic or latino</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>417.0</td>\n",
              "      <td>26627.0</td>\n",
              "      <td>None</td>\n",
              "      <td>Infiltrating duct carcinoma, NOS</td>\n",
              "      <td>8500/3</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>TCGA-E2-A156</td>\n",
              "      <td>TCGA-BRCA</td>\n",
              "      <td>BRCA</td>\n",
              "      <td>female</td>\n",
              "      <td>white</td>\n",
              "      <td>not hispanic or latino</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>726.0</td>\n",
              "      <td>22422.0</td>\n",
              "      <td>None</td>\n",
              "      <td>Infiltrating duct carcinoma, NOS</td>\n",
              "      <td>8500/3</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>TCGA-A7-A56D</td>\n",
              "      <td>TCGA-BRCA</td>\n",
              "      <td>BRCA</td>\n",
              "      <td>female</td>\n",
              "      <td>black or african american</td>\n",
              "      <td>not hispanic or latino</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>448.0</td>\n",
              "      <td>30783.0</td>\n",
              "      <td>None</td>\n",
              "      <td>Infiltrating duct carcinoma, NOS</td>\n",
              "      <td>8500/3</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>TCGA-AR-A0U1</td>\n",
              "      <td>TCGA-BRCA</td>\n",
              "      <td>BRCA</td>\n",
              "      <td>female</td>\n",
              "      <td>white</td>\n",
              "      <td>not hispanic or latino</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>4052.0</td>\n",
              "      <td>13467.0</td>\n",
              "      <td>None</td>\n",
              "      <td>Infiltrating duct carcinoma, NOS</td>\n",
              "      <td>8500/3</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-be5db493-b001-4acf-8280-36915682f7b6')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-be5db493-b001-4acf-8280-36915682f7b6 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-be5db493-b001-4acf-8280-36915682f7b6');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-f5ff9891-f9b9-4456-8e80-206107bdc70d\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-f5ff9891-f9b9-4456-8e80-206107bdc70d')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-f5ff9891-f9b9-4456-8e80-206107bdc70d button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "clinical_df",
              "repr_error": "Out of range float values are not JSON compliant: nan"
            }
          },
          "metadata": {},
          "execution_count": 7
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 13) Align clinical table to expression samples and save to PROCESSED_DIR\n",
        "\n",
        "# Load your main log-TPM aligned matrix (samples x genes)\n",
        "EXPR_ALIGNED_PATH = PROCESSED_DIR / \"tcga_tpm_like_log10_bulkrnabert_aligned.parquet\"\n",
        "expr_aligned = pd.read_parquet(EXPR_ALIGNED_PATH)\n",
        "print(\"Expression aligned shape:\", expr_aligned.shape)\n",
        "\n",
        "# Align by sample_id\n",
        "clinical_df = clinical_df.set_index(\"sample_id\")\n",
        "\n",
        "common_samples = expr_aligned.index.intersection(clinical_df.index)\n",
        "print(\"Samples in expression:\", expr_aligned.shape[0])\n",
        "print(\"Samples in clinical  :\", clinical_df.shape[0])\n",
        "print(\"Common samples       :\", len(common_samples))\n",
        "\n",
        "clinical_aligned = clinical_df.loc[common_samples].copy()\n",
        "\n",
        "# Basic diagnostics\n",
        "print(\"\\nNon-null survival_time:\", clinical_aligned[\"survival_time\"].notna().sum())\n",
        "print(\"Non-null event        :\", clinical_aligned[\"event\"].notna().sum())\n",
        "print(\"\\nCohort counts:\")\n",
        "print(clinical_aligned[\"cohort\"].value_counts())\n",
        "\n",
        "# Save cleaned, aligned clinical table\n",
        "CLIN_OUT = PROCESSED_DIR / \"tcga_clinical_bulkrnabert_cohorts.parquet\"\n",
        "clinical_aligned.to_parquet(CLIN_OUT)\n",
        "\n",
        "print(\"\\nSaved aligned clinical to:\", CLIN_OUT)\n"
      ],
      "metadata": {
        "id": "i94-rhMv4iD8",
        "outputId": "0f954ff7-29d4-4750-c066-1b5a7cc4bd9e",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "id": "i94-rhMv4iD8",
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Expression aligned shape: (3385, 19062)\n",
            "Samples in expression: 3385\n",
            "Samples in clinical  : 3788\n",
            "Common samples       : 3385\n",
            "\n",
            "Non-null survival_time: 0\n",
            "Non-null event        : 0\n",
            "\n",
            "Cohort counts:\n",
            "cohort\n",
            "BRCA    1095\n",
            "UCEC     557\n",
            "LUAD     518\n",
            "LGG      516\n",
            "BLCA     406\n",
            "GBM      293\n",
            "Name: count, dtype: int64\n",
            "\n",
            "Saved aligned clinical to: /content/drive/MyDrive/bdh_challenge_2025_data/processed/tcga_clinical_bulkrnabert_cohorts.parquet\n"
          ]
        }
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python",
      "version": "3.10"
    },
    "colab": {
      "provenance": [],
      "toc_visible": true,
      "machine_shape": "hm"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}